src/Server/Services/Analytics/AnalyticsService.lua

```diff
@@ -265,155 +313,356 @@ export type RemoteConfig = {
+
+	MarketplaceService.PromptProductPurchaseFinished:Connect(
+		function(userId: number, productId: number, purchased: boolean)
 			if not purchased then
 				return
 			end
-			
+
+			local productInfo = self:_getProductInfo(productId, Enum.InfoType.Product)
+
 			self:LogMarketplacePurchase({
-				userId = player.UserId,
-				itemType = "Asset",
-				id = assetId,
+				userId = userId,
+				itemType = "Product",
+				id = productInfo and productInfo.Name or productId,
 				cartType = "PromptPurchase",
+				robuxPrice = productInfo and productInfo.PriceInRobux,
 			})
 		end
 	)
-	
+
+	MarketplaceService.PromptPurchaseFinished:Connect(function(player: Player, assetId: number, purchased: boolean)
+		if not purchased then
+			return
+		end
+
+		local productInfo = self:_getProductInfo(assetId, Enum.InfoType.Asset)
+
+		self:LogMarketplacePurchase({
+			userId = player.UserId,
+			itemType = "Asset",
+			id = productInfo and productInfo.Name or assetId,
+			cartType = "PromptPurchase",
+			robuxPrice = productInfo and productInfo.PriceInRobux,
+		})
+	end)
+
 	MarketplaceService.PromptSubscriptionPurchaseFinished:Connect(
 		function(player: Player, subscriptionId: string, didTryPurchasing: boolean)
 			if not didTryPurchasing then
 				return
 			end
-			
+
+			local success, result = pcall(function()
+				local cacheIndex: string = `subscription-{subscriptionId}`
+
+				if self._cache[cacheIndex] then
+					return self._cache[cacheIndex]
+				end
+
+				local productInfo = MarketplaceService:GetSubscriptionProductInfoAsync(subscriptionId)
+
+				self._cache[cacheIndex] = productInfo
+
+				return productInfo
+			end)
+
+			local robuxPrice = 0
+
+			if success then
+				local priceNumberString = string.match(result.DisplayPrice, "%d+%.?%d*")
+				local dollarPrice = tonumber(priceNumberString or "")
+
+				if dollarPrice ~= nil then
+					-- Convert from USD to Robux using GA's conversion rate
+					robuxPrice = ((dollarPrice * 100) / 0.7) / 0.35
+				end
+			end
+
 			self:LogMarketplacePurchase({
 				userId = player.UserId,
 				itemType = "Subscription",
-				id = subscriptionId,
+				id = success and result.Name or subscriptionId,
+				amount = robuxPrice,
 				cartType = "PromptPurchase",
 			})
 		end
 	)
-	
+
+	Players.PlayerAdded:Connect(function(player: Player)
+		if player.FollowUserId ~= 0 then
+			self:LogPlayerEvent({
+				userId = player.UserId,
+				event = "Player:JoinedFriend",
+			})
+		end
+	end)
+
 	Players.PlayerRemoving:Connect(function(player: Player)
 		self:_flushTrackedEvents(player)
 	end)
-	
+
+	return
+end
+
+--[=[
+	@private
+	@param fn function
+	@param ... any
+]=]
+function AnalyticsService:_wrapper(fn: string, ...): nil
+	local args = { ... }
+
+	Promise.new(function(resolve, reject)
+		local success, result = pcall(GameAnalytics[fn], GameAnalytics, unpack(args))
+
+		if success then
+			return resolve(result)
+		else
+			return reject(result)
+		end
+	end):catch(function(err)
+		warn(err)
+	end)
```

Comment:
```suggestion
	local promiseFunction = Promise.promisify(GameAnalytics[fn])
	promiseFunction(GameAnalytics,unpack(args)):catch(warn)
```

Promisify (https://eryn.io/roblox-lua-promise/api/Promise#promisify) allows you to take a regular function and convert it into a promise. If the provided function errors, the returned promise rejects. If the provided function does not error, the returned promise resolves with the callback from the provided function.